packages/python/plotly/plotly/data/__init__.py

```diff
@@ -128,95 +196,221 @@ def election_geojson():
         `['nation', 'gold', 'silver', 'bronze']`.
         If `indexed` is True, the 'nation' column is used as the index and the column index
-        is named 'medal'"""
-    df = _get_dataset("medals")
-    if indexed:
-        df = df.set_index("nation")
+        is named 'medal'
+    """
+
+    if indexed and return_type != "pandas":
+        msg = "Cannot set index for backend different from pandas"
+        raise NotImplementedError(msg)
+
+    df = nw.from_native(
+        _get_dataset("medals", return_type=return_type), eager_only=True
+    )
+    if indexed:  # then it must be pandas
+        df = df.to_native().set_index("nation")
         df.columns.name = "medal"
-    return df
+        return df
+    return df.to_native()
 
 
-def medals_long(indexed=False):
+def medals_long(indexed=False, return_type="pandas"):
     """
     This dataset represents the medal table for Olympic Short Track Speed Skating for the
     top three nations as of 2020.
 
-    Returns:
-        A `pandas.DataFrame` with 9 rows and the following columns:
-        `['nation', 'medal', 'count']`.
-        If `indexed` is True, the 'nation' column is used as the index."""
-    df = _get_dataset("medals").melt(
-        id_vars=["nation"], value_name="count", var_name="medal"
+    Parameters
+    ----------
+    indexed: bool
+        Whether or not the 'nation' column is used as the index.
+        Applicable only if `return_type='pandas'`
+
+    return_type: {'pandas', 'polars', 'pyarrow'}
+        Type of the resulting dataframe
+
+    Returns
+    -------
+    Dataframe of `return_type` type
+        Dataframe with 9 rows and the following columns: `['nation', 'medal', 'count']`.
+        If `indexed` is True, the 'nation' column is used as the index.
+    """
+
+    if indexed and return_type != "pandas":
+        msg = "Cannot set index for backend different from pandas"
+        raise NotImplementedError(msg)
+
+    df = nw.from_native(
+        _get_dataset("medals", return_type=return_type), eager_only=True
+    ).unpivot(
+        index=["nation"],
+        value_name="count",
+        variable_name="medal",
     )
     if indexed:
-        df = df.set_index("nation")
-    return df
+        df = nw.maybe_set_index(df, "nation")
+    return df.to_native()
+
+
+def _get_dataset(d, return_type):
+    """
+    Loads the dataset using the specified backend.
+
+    Notice that the available backends are 'pandas', 'polars', 'pyarrow' and they all
+    have a `read_csv` function. Therefore we can dynamically load the library via
+    `importlib.import_module` and then call `backend.read_csv(filepath)`.
 
+    Parameters
+    ----------
+    d: str
+        Name of the dataset to load.
 
-def _get_dataset(d):
-    import pandas
+    return_type: {'pandas', 'polars', 'pyarrow'}
+        Type of the resulting dataframe
+
+    Returns
+    -------
+    Dataframe of `return_type` type
+    """
     import os
+    from importlib import import_module
```

Comment:
@archmoj @LiamConnors Is there some performance / other reason you can think of why the imports in this file are all nested inside the individual functions rather than being at the top? If not, I'd prefer to move them to the top.

@FBruzzesi I'd prefer `AVAILABLE_BACKENDS` to be defined at the top as well.

Reply:
Moved imports and `AVAILABLE_BACKENDS` to top of the file with the last commit